### PR TITLE
GTC: GridTools C++ backend

### PR DESCRIPTION
**Pull request by havogt**
Adds the GridTools C++ 2.x IR and codegenerator.

### CODE REVIEW

File: tests/test_unittest/test_gtc/test_gtcpp_compilation.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            GTFunctorBuilder("fun").add_accessors(accessors).add_apply_method(apply_method).build()
+        )
+        .build()
+    )
+
+
+@pytest.mark.parametrize(
+    "apply_method,expected_regex",
+    [
+        (GTApplyMethodBuilder().build(), r""),
```
Comment (DropD) [inline tests/test_unittest/test_gtc/test_gtcpp_compilation.py:136]: expected regex is empty

File: tests/test_unittest/test_gtc/test_gtcpp_compilation.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        (
+            GTApplyMethodBuilder()
+            .add_stmt(
+                AssignStmt(
+                    left=AccessorRefBuilder("a").build(),
+                    right=AccessorRefBuilder(name="b").build(),
+                )
+            )
+            .build(),
+            r"",
```
Comment (DropD) [inline tests/test_unittest/test_gtc/test_gtcpp_compilation.py:146]: expected regex is empty

File: tests/test_unittest/test_gtc/test_gtcpp_compilation.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    If it matches `True` is returned.
+    If it doesn't match an `AssertionError` is raised.
+    """
+    assert re.search(regexp, str(value)), "Pattern {!r} does not match {!r}".format(
+        regexp, str(value)
+    )
+    return True
+
+
+# TODO can I get tmp_path automagically here?
```
Comment (DropD) [inline tests/test_unittest/test_gtc/test_gtcpp_compilation.py:35]: yes, by making it a fixture that returns the function

File: tests/test_unittest/test_gtc/gtcpp_utils.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    def build(self) -> GTInterval:
+        return GTInterval(from_level=self._from_level, to_level=self._to_level)
+
+
+class GTApplyMethodBuilder:
+    def __init__(self) -> None:
+        self._interval = GTIntervalBuilder().build()
+        self._body = []
+
+    def add_stmt(self, stmt: Stmt):
```
Comment (DropD) [inline tests/test_unittest/test_gtc/gtcpp_utils.py:56]: From here on most self-return type annotations are missing.

Comment (havogt) [reply tests/test_unittest/test_gtc/gtcpp_utils.py:56]: done

File: tests/test_unittest/test_gtc/test_gtcpp_compilation.py (hunk 1 above)
Comment (havogt) [reply tests/test_unittest/test_gtc/test_gtcpp_compilation.py:136]: done

File: tests/test_unittest/test_gtc/test_gtcpp_compilation.py (hunk 2 above)
Comment (havogt) [reply tests/test_unittest/test_gtc/test_gtcpp_compilation.py:146]: done

File: tests/test_unittest/test_gtc/test_gtcpp_compilation.py (hunk 3 above)
Comment (havogt) [reply tests/test_unittest/test_gtc/test_gtcpp_compilation.py]: left it for now as the indirection doesn't simplify the code

**Review by DropD — APPROVED** [review]

**Pull request merged**